# Proyecto integral

Este proyecto tiene como propósito identificar patrones que determinen si un videojuego tiene éxito o no. Esto para detectar proyectos prometedores y planificar campañas publicitarias.

El proyecto constará de 5 fases.
1) **Cargar los datos** y analizar la información general,
2) **transformar/preparar** los datos para
3) **analizar** los datos con el objetivo de 
4) **crear un perfil de usuario** para cada región.
5) **Prueba de hipótesis** respecto a las calificaciones promedio de los usuarios clasificando por plataformas y género.
6) **Conclusiones generales**.

## Fase 1. Carga de datos y analisis general

Primero cargaremos los datos usando pandas y después daremos una visualización general del DataFrame así como una descrición de las variables y posibles problemas estructurales.

In [1]:
import pandas as pd

In [2]:
juegos = pd.read_csv('datasets/games.csv')

In [3]:
juegos.head(10)

,Name,Platform,Year_of_Release,Genre,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score,User_Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
5,Tetris,GB,1989.0,Puzzle,23.20,2.26,4.22,0.58,NaN,NaN,NaN
6,New Super Mario Bros.,DS,2006.0,Platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,Wii Play,Wii,2006.0,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E
8,New Super Mario Bros. Wii,Wii,2009.0,Platform,14.44,6.94,4.70,2.24,87.0,8.4,E
9,Duck Hunt,NES,1984.0,Shooter,26.93,0.63,0.28,0.47,NaN,NaN,NaN


In [4]:
juegos.describe()

,Year_of_Release,NA_sales,EU_sales,JP_sales,Other_sales,Critic_Score
count,16446.000000,16715.000000,16715.000000,16715.000000,16715.000000,8137.000000
mean,2006.484616,0.263377,0.145060,0.077617,0.047342,68.967679
std,5.877050,0.813604,0.503339,0.308853,0.186731,13.938165
min,1980.000000,0.000000,0.000000,0.000000,0.000000,13.000000
25%,2003.000000,0.000000,0.000000,0.000000,0.000000,60.000000
50%,2007.000000,0.080000,0.020000,0.000000,0.010000,71.000000
75%,2010.000000,0.240000,0.110000,0.040000,0.030000,79.000000
max,2016.000000,41.360000,28.960000,10.220000,10.570000,98.000000


En el dataframe de <code>juegos</code> podemos ver que hay 11 columnas
1) **Name**: nombre del videojuego.
2) **Plataform**: plataforma donde se encuentra el videojuego.
3) **Year_of_Realise**: año de lanzamiento.
4) **Genre**: género del videojuego.
5) **NA_sales**: Ventas en Norteamérica en millones de dólares estadounidenses
6) **EU_sales**: Ventas en Europa en millones de dólares estadounidenses
7) **JP_sales**: Ventas en Japón en millones de dólares estadounidenses
8) **Other_sales**: Ventas en otro paises en millones de dolares estadounidenses
9) **Critic_Score**: Valoración por críticos (máximo de 100)
10) **User_score**: Valoración de usuarios (máximo de 10)
11) **Rating**: Ranking de la Junta de clasificación de software de entretenimiento, *Entertainment Software Rating Board* (ESRB). Evalúa el contenido de un juego y asigna una clasificación de edad como Adolescente o Adulto.

### **Clasificación ESRB**

La **ESRB** (*Entertainment Software Rating Board*) es una organización que clasifica los videojuegos según la edad recomendada y el tipo de contenido. Estas categorías ayudan a identificar el público objetivo de cada juego.

Las clasificaciones más comunes son:

- **E (Everyone)**: contenido apto para todo público.
- **E10+ (Everyone 10+)**: contenido apto para mayores de 10 años.
- **T (Teen)**: contenido apto para adolescentes de 13 años en adelante.
- **M (Mature 17+)**: contenido apto para mayores de 17 años.
- **AO (Adults Only 18+)**: contenido solo para adultos.
- **EC (Early Childhood)**: contenido dirigido a niños pequeños.
- **RP (Rating Pending)**: clasificación aún no asignada oficialmente.

Esta variable podría ser útil en el análisis para estudiar si la clasificación por edad influye en las ventas de los videojuegos en distintas regiones.

## Fase 2.Transformación y preparación de los datos

En esta fase se trabajará el dataframe de juegos con el propósito de preparar el dataset para el análisis posterior y se realizará de la siguiente manera:
1) Remplazar el nombre de las columnas poniendo todo en minúsculas.
2) Ingeniería de características
   1) Convertir los tipos de datos en los necesarios y/o adecuados para su tratamiento o naturaleza.
   2) Buscar duplicados, analizar su naturaleza y decidir si eliminarlos o trasformarlos.
   3) Buscar los datos faltantes para analizar su naturaleza y decidir cómo transformalos o eliminarlos de ser necesario.
3) Se calculará las ventas totales y se agregará la columna al final de la tabla/dataframe de juegos.

Se describirán todo los pasos y decisiones a detalle.

### Nombre de la columnas

In [5]:
columnas = juegos.columns
print(columnas)

Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'NA_sales', 'EU_sales',
       'JP_sales', 'Other_sales', 'Critic_Score', 'User_Score', 'Rating'],
      dtype='object')


In [9]:
columnas_nuevas = []
for columna in juegos.columns:
    columna = columna.lower().strip()
    columnas_nuevas.append(columna)

juegos.columns = columnas_nuevas
juegos.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

In [10]:
juegos.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


### Ingeniería de características

Después de haber cambiado los títulos de las columnas, con el fin de evitar confusiones y mantener una estructura uniforme al llamar a las columnas en el futuro; se continua con el analisis de los tipos de datos de cada columna, datos faltantes y duplicados.

Primero veremos la información general del dataframe

In [11]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16715 non-null  object 
 2   year_of_release  16446 non-null  float64
 3   genre            16713 non-null  object 
 4   na_sales         16715 non-null  float64
 5   eu_sales         16715 non-null  float64
 6   jp_sales         16715 non-null  float64
 7   other_sales      16715 non-null  float64
 8   critic_score     8137 non-null   float64
 9   user_score       10014 non-null  object 
 10  rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


Podemos ver que el dataframe continiene 11 columnas con 16715 entradas/filas. 
- **name**: tipo de dato object. Será necesario comprobar que todos los datos sean str o convertirlos a tal. También analizar esos valores nulos mostrados.
- **plataform**: tipo de dato object. Será necesario comprobar que todos los datos sean str o convertirlos a tal. 
- **year_of_realise**: tipo de dato float. Será necesario convertirlo a int. Se analizarán los valores nulos.
- **genre**: tipo de dato object. Será necesario comprobar que todos los datos sean str o convertirlos a tal. Se analizarán los valores nulos
- **na_sales**: tipo de dato float. No es necesario convertirlo.
- **eu_sales**: tipo de dato float. No es necesario convertirlo.
- **jp_sales**: tipo de dato float. No es necesario convertirlo.
- **other_sales**: tipo de dato float. No es necesario convertirlo.
- **critic_score**: tipo de dato float. No es necesario convertirlo. Se analizarán los valores nulos.
- **user_score**: tipo de dato object. Será necesario convertirlo a float. Se analizarán los valores nulos.
- **rating**: tipo de dato object. Será necesario comprobar que todos los datos sean str o convertirlos a tal. Se analizarán los valores nulos.

Primero veamos a detalle la variación de tipo de datos en las columnas.

In [12]:
for i in juegos.columns:
    print('------------')
    print(juegos[i].apply(type).value_counts())
    print('------------')

------------
name
<class 'str'>      16713
<class 'float'>        2
Name: count, dtype: int64
------------
------------
platform
<class 'str'>    16715
Name: count, dtype: int64
------------
------------
year_of_release
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
genre
<class 'str'>      16713
<class 'float'>        2
Name: count, dtype: int64
------------
------------
na_sales
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
eu_sales
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
jp_sales
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
other_sales
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
critic_score
<class 'float'>    16715
Name: count, dtype: int64
------------
------------
user_score
<class 'str'>      10014
<class 'float'>     6701
Name: count, dtype: int64
------------
------------
rating
<class 'str'>      9949
<class 'float'

Después de contabilizar el tipo de valor en cada columna tenemos
- **name**: Todos los valores no nulos están correctamente etiquetados como cadena de texto. Habrá que analizar los valores nulos.
- **plataform**: Todos los valores no nulos están correctamente etiquetados como cadena de texto.
- **year_of_realise**: Todos los valores no nulos están etiquetados como números decimales. Se convertirá a entero y se analizarán los valores nulos.
- **genre**: Todos los valores no nulos están correctamente etiquetados como cadena de texto. Habrá que analizar los valores nulos.
- **na_sales**: Todos los valores no nulos están etiquetados como números decimales. No se harán modificaciones.
- **eu_sales**: Todos los valores no nulos están etiquetados como números decimales. No se harán modificaciones.
- **jp_sales**: Todos los valores no nulos están etiquetados como números decimales. No se harán modificaciones.
- **other_sales**: Todos los valores no nulos están etiquetados como números decimales. No se harán modificaciones.
- **critic_score**: Todos los valores no nulos están etiquetados como números decimales. No se harán modificaciones. No hay valores nulos.
- **user_score**: Todos los valores no nulos están etiquetados como cadena de texto. Se convertirán a números decimales. Los valores de float corresponden a los valores nulos. Se analizará si es necesario eliminar o transformar esos valores nulos.
- **rating**: Todos los valores no nulos están etiquetados como string, habrá que convertilos a decimales y analizar los valores nulos para decidir si eliminarlos o transformarlos.

#### Conversión de datos

In [13]:
# Conversión de valores de year_of_realise de float a int
juegos['year_of_release'] = juegos['year_of_release'].astype('Int64')

In [14]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16715 non-null  object 
 2   year_of_release  16446 non-null  Int64  
 3   genre            16713 non-null  object 
 4   na_sales         16715 non-null  float64
 5   eu_sales         16715 non-null  float64
 6   jp_sales         16715 non-null  float64
 7   other_sales      16715 non-null  float64
 8   critic_score     8137 non-null   float64
 9   user_score       10014 non-null  object 
 10  rating           9949 non-null   object 
dtypes: Int64(1), float64(5), object(5)
memory usage: 1.4+ MB


Veamos el caso especial de user_score y veamos los valores únicos de la columna

In [17]:
# valores únicos de la columna user_score
juegos['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

Al ver que hay valores nulos, cadenas de texto que pueden presentarse como float y cadenas de texto <code>tbd</code>, <i>to be determined</i>, se decide convertir los valores de user_score a float y los valores nulos a NaN. Después se analizará si es necesario eliminar o transformar esos valores nulos.

In [18]:
# conversión del tipo de dato de user_score
juegos['user_score'] = pd.to_numeric(juegos['user_score'], errors='coerce')

In [19]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16715 non-null  object 
 2   year_of_release  16446 non-null  Int64  
 3   genre            16713 non-null  object 
 4   na_sales         16715 non-null  float64
 5   eu_sales         16715 non-null  float64
 6   jp_sales         16715 non-null  float64
 7   other_sales      16715 non-null  float64
 8   critic_score     8137 non-null   float64
 9   user_score       7590 non-null   float64
 10  rating           9949 non-null   object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


Vamos ahora la columna de rating

In [20]:
# Valores únicos de rating
juegos['rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

***Clasificaciones ESRB encontradas***

En la columna `rating` se observan las principales categorías de la ESRB, que indican la edad recomendada y el tipo de contenido del videojuego:

- **E (Everyone):** apto para todo público.  
- **E10+ (Everyone 10+):** recomendado para mayores de 10 años.  
- **T (Teen):** contenido para adolescentes (13+).  
- **M (Mature 17+):** contenido para mayores de 17 años.  
- **AO (Adults Only 18+):** solo para adultos (18+).  
- **EC (Early Childhood):** dirigido a niños pequeños.  
- **RP (Rating Pending):** clasificación pendiente.  
- **K-A (Kids to Adults):** clasificación antigua, equivalente aproximada a **E**.

La categoría K-A se agrupará con E para evitar confunsiones. Por otro lado, los valores nulos se tratarán un poco más adelante.

In [22]:
# Cambio de K-A a E
juegos['rating'] = juegos.rating.replace('K-A', 'E')
juegos['rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'AO', 'EC', 'RP'], dtype=object)

#### Valores auscentes

Ya hemos convertido el tipo de dato al adecuado, ahora se analizarán los valores nulos.

In [23]:
juegos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16715 non-null  object 
 2   year_of_release  16446 non-null  Int64  
 3   genre            16713 non-null  object 
 4   na_sales         16715 non-null  float64
 5   eu_sales         16715 non-null  float64
 6   jp_sales         16715 non-null  float64
 7   other_sales      16715 non-null  float64
 8   critic_score     8137 non-null   float64
 9   user_score       7590 non-null   float64
 10  rating           9949 non-null   object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.4+ MB


#### Duplicados

### Columna de totales

## Fase 3. Analisis de datos


En esta parte se realizará un análisis exploratorio de los datos para identificar patrones y tendencias. Se estudiará:

1) Juegos lanzados en diferentes años.
2) Comportamiento en las plataformas.
   1) Variación de ventas.
   2) Ciclo de vida de las distintas plataformas.
3) Seleccionar intervalos de tiempo relevantes.

Después de seleccionar los intervalos de tiempo relevantes se estudiará:

4) Plataformas líderes en ventas. Crecimiento y decreciemiento de plataformas.
5) Dispersión de ventas globales por plataforma.
6) Relación entre las opiniones de los usuarios y las ventas de la plataforma.
7) Comparación de videojuegos por plataforma.

Finalmente se buscará responder también a las preguntas ¿qué géneros son más rentables? y ¿hay características determinantes entre los géneros con más ventas y los géneros con menos ventas?

### Juegos lanzados en diferentes años

¿son significativos los datos de cada periodo?

### Comportamiento en las plataformas

#### Variación de ventas

Distribución de ventas totales de cada año en función de la plataformas con mayor ventas.

#### Ciclo de vida de las distintas plataformas

[Busca las plataformas que solían ser populares pero que ahora no tienen ventas. ¿Cuánto tardan generalmente las nuevas plataformas en aparecer y las antiguas en desaparecer?]

### Intervalos de tiempo relevantes

### Plataformas líderes

[¿Qué plataformas son líderes en ventas? ¿Cuáles crecen y cuáles se reducen? Elige varias plataformas potencialmente rentables.]

### Dispersión de ventas globales por plataforma

[Crea un diagrama de caja para las ventas globales de todos los juegos, desglosados por plataforma. ¿Son significativas las diferencias en las ventas? ¿Qué sucede con las ventas promedio en varias plataformas? Describe tus hallazgos.]

### Relación entre las opiniones de los usuarios y las ventas de la plataforma

[Mira cómo las reseñas de usuarios y profesionales afectan las ventas de una plataforma popular (tu elección). Crea un gráfico de dispersión y calcula la correlación entre las reseñas y las ventas. Saca conclusiones.]

### Comparación de videojuegos por plataforma

### Conclusiones generales del análisis de datos

[¿Qué se puede decir de los géneros más rentables? ¿Puedes generalizar acerca de los géneros con ventas altas y bajas?]

## Fase 4. Perfil de usuario por región.


Para esta fase tendremos 3 secciones relevantes:
1) Se describirán las variaciones de las cuotas de mercado entre regiones, basándose en las 5 plataformas principales.
2) Se analizará la diferencia entre los 5 géneros principales.
3) Se analizará si las clasificaciones de ESRB afectan las ventas en cada región.

## Fase 5. Pruebas de hipótesis respecto a las calificaciones promedio de los usuarios clasificando por plataformas y género.

Ahora comprobaremos algunas hipótesis respecto a las calificaciones promedio de los usuarios clasificando por plataformas y género.
- Las calificaciones promedio de los usuarios para las plataformas Xbox One y PC son las mismas.
- Las calificaciones promedio de los usuarios para los géneros de Acción y Deportes son diferentes. 

## Fase 6. Conclusiones generales